FLT/Mathlib/Topology/Algebra/Group/Units.lean

```diff
@@ -7,9 +7,88 @@ lemma Submonoid.units_isOpen {M : Type*} [TopologicalSpace M] [Monoid M]
   {U : Submonoid M} (hU : IsOpen (U : Set M)) : IsOpen (U.units : Set Mˣ) :=
   (hU.preimage Units.continuous_val).inter (hU.preimage Units.continuous_coe_inv)
 
+ -- FLT#588
 lemma Submonoid.units_isCompact {M : Type*} [TopologicalSpace M] [Monoid M] [ContinuousMul M]
     [T2Space M] {U : Submonoid M} (hU : IsCompact (U : Set M)) : IsCompact (U.units : Set Mˣ) := by
-  sorry -- FLT#588
+  let φ : Mˣ → M × M := fun x ↦  (↑x, ↑x⁻¹)
+  let p : M × M → M := fun (u,v) => u * v
+  let q : M × M → M := fun (u,v) => v * u
+
+  have range_φ : p ⁻¹' {1} ∩ q ⁻¹' {1} = Set.range φ := by
+    apply Set.ext
+    intro (u,v)
+    simp_all only[p,q,φ]
+    simp_all only [Set.image_univ, Set.mem_range, Prod.mk.injEq,
+         Set.mem_inter_iff, Set.mem_preimage, Set.mem_singleton_iff]
+    constructor
+    · -- Forward direction: u * v = 1 ∧ v * u = 1 ⇒ (u, v) ∈ range φ
+      intro h
+      let y : Mˣ := ⟨u, v, h.1, h.2⟩
+      use y
+      simp only [Units.inv_mk, and_self, y]
+    · -- Backward direction: (u, v) ∈ Set.range φ ⇒ (u * v = 1 ∧ v * u = 1)
+      intro h
+      obtain ⟨w, h⟩ := h
+      obtain ⟨left, right⟩ := h
+      subst left right
+      simp_all only [Units.mul_inv, Units.inv_mul, and_self, q, φ, p]
+
+  have closed_range_φ : IsClosed (Set.range φ):= by
+    rw [← range_φ]
+    exact IsClosed.inter
+      (isClosed_singleton.preimage continuous_mul)
+      (isClosed_singleton.preimage  (continuous_mul.comp continuous_swap))
+  have isCompacthUU : IsCompact ((U : Set M) ×ˢ (U : Set M)) := hU.prod hU
+
+  have φUu_subset : φ '' (U.units : Set Mˣ) ⊆ Set.range φ ∩ U ×ˢ U := by
+    rintro z ⟨x, hxU, rfl⟩
+    constructor
+    · -- φ x ∈ Set.range φ
+      simp_all only [SetLike.mem_coe, Set.mem_range, Prod.mk.injEq, p, q, φ]
+      apply Exists.intro
+      · apply And.intro
+        on_goal 2 => {rfl
+        }
+        · simp_all only [p, q, φ]
+    · -- φ x ∈ ↑U ×ˢ ↑U
+      simp_all only [SetLike.mem_coe]
+      exact hxU
```

Comment:
```suggestion
    · -- φ x ∈ ↑U ×ˢ ↑U
      exact hxU
```
The `simp_all only` is only changing things up to definitional equality, and `exact` works up to definitional equality so just sees right through it.

So in fact here's a complete proof of `φUu_subset`:
```lean
  have φUu_subset : φ '' (U.units : Set Mˣ) ⊆ Set.range φ ∩ U ×ˢ U := by
    rintro z ⟨x, hxU, rfl⟩
    exact ⟨⟨x, rfl⟩, hxU⟩ 
```